File: algorithms-data-structures/ts/data-structures/stacks/min-stack/min-stack.ts
export class MinStack {
  stack: number[];
  minStack: number[];
  size: number;

  constructor() {
    this.stack = [];
    this.minStack = [];
    this.size = 0;
  }

  push(n: number): void {
    if (this.size === 0) {
      this.stack.push(n);
      this.minStack.push(n);
    } else {
      let currentMin = this.minStack.at(-1)!;
      if (n < currentMin) {
        this.minStack.push(n);
      } else {
        this.minStack.push(currentMin);
      }
      this.stack.push(n);
    }
    this.size++;
  }
  pop(): void {
    this.stack.pop();
    this.minStack.pop();
    this.size--;
  }

  top(): number {
    return this.stack.at(-1)!;
  }
  getMin(): number {
    return this.minStack.at(-1)!;
  }
}

File: algorithms-data-structures/go/src/data-structures/stacks/min-stack/min-stack.go
package minstack

type MinStack struct {
	stack []int
	mins  []int
}

func Constructor() MinStack {
	return MinStack{}
}

func (ms *MinStack) Push(val int) {
	if len(ms.stack) == 0 {
		ms.stack = append(ms.stack, val)
		ms.mins = append(ms.mins, val)
	} else {
		top := ms.mins[len(ms.mins)-1]
		if val < top {
			ms.mins = append(ms.mins, val)
		} else {
			ms.mins = append(ms.mins, top)
		}
		ms.stack = append(ms.stack, val)
	}

}

func (ms *MinStack) Pop() {
	if len(ms.stack) > 0 {
		ms.stack = ms.stack[:len(ms.stack)-1]
		ms.mins = ms.mins[:len(ms.mins)-1]
	}
}

func (ms *MinStack) Top() int {
	if len(ms.stack) > 0 {
		return ms.stack[len(ms.stack)-1]
	}
	return -1
}

func (ms *MinStack) GetMin() int {
	if len(ms.mins) == 0 {
		return -1
	}
	return ms.mins[len(ms.mins)-1]
}
